Kentico/xperience-algolia
# Feat/distinct index

### Motivation

Implements #19 

### Checklist


### How to test

Register an index using the new `DistinctOptions` parameter:

```cs
Service.Use<IAlgoliaIndexRegister>(new DefaultAlgoliaIndexRegister()
                .Add<SiteSearchModel>(SiteSearchModel.IndexName, distinctOptions: new DistinctOptions(nameof(SiteSearchModel.DocumentName), 1))
```

In the search model, implement the `AlgoliaSearchModel.SplitData` method to generate multiple `JObject`s with the desired data. For example, split a large "Content" value by paragraph:

```cs
public override IEnumerable<JObject> SplitData(JObject originalData)
{
    var originalId = originalData.Value<string>("objectID");
    var content = originalData.Value<string>(nameof(Content));
    System.Text.RegularExpressions.Match m = System.Text.RegularExpressions.Regex.Match(content, @"<p>\s*(.+?)\s*</p>");
    List<string> paragraphs = new List<string>();
    while (m.Success)
    {
        paragraphs.Add(m.Value);
        m = m.NextMatch();
    }

    return paragraphs.Select((p, index) => {
        var data = (JObject)originalData.DeepClone();
        data["objectID"] = $"{originalId}-{index}";
        data[nameof(Content)] = p;
        return data;
    });
}
```

On page creation/update, multiple records should be created within Algolia containing only a single paragraph. On page deletion, all "chunks" in Algolia are deleted. While searching on the front end, only the most relevant "chunk" should be returned (when using `DistinctLevel` of 1).
